src/Raven.Server/ServerWide/CompareExchangeExpirationStorage.cs

```diff
@@ -0,0 +1,165 @@
+        {
+            if (value.TryGetMember(Constants.Documents.Metadata.Key, out var metadata))
+            {
+                if (metadata is BlittableJsonReaderObject bjro && bjro.TryGet(Constants.Documents.Metadata.Expires, out object obj))
+                {
+                    if (obj is LazyStringValue expirationDate)
+                    {
+                        if (DateTime.TryParseExact(expirationDate, DefaultFormat.DateTimeFormatsToRead, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out DateTime date) == false)
+                        {
+                            var inner = new InvalidOperationException(
+                                $"The expiration date format for compare exchange '{CompareExchangeKey.SplitStorageKey(storageKey).Key}' is not valid: '{expirationDate}'. Use the following format: {DateTime.UtcNow:O}");
+                            throw new RachisApplyException("Could not apply command.", inner);
+                        }
+                        var expiry = date.ToUniversalTime();
+                        ticks = expiry.Ticks;
+                        return true;
+                    }
+                    else
+                    {
+                        var inner = new InvalidOperationException($"The type of {Constants.Documents.Metadata.Expires} metadata for compare exchange '{CompareExchangeKey.SplitStorageKey(storageKey).Key}' is not valid. Use the following type: {nameof(DateTime)}");
+                        throw new RachisApplyException("Could not apply command.", inner);
+                    }
+                }
+            }
+
+            ticks = default;
+            return false;
+        }
+
+        public static unsafe bool DeleteExpiredCompareExchange(ClusterOperationContext context, Table items, long ticks, long take = long.MaxValue)
+        {
+            var expired = new Dictionary<Slice, List<Slice>>();
+
+            foreach ((Slice keySlice, long expiredTicks, Slice ticksSlice) in GetExpiredValues(context, ticks))
+            {
+                if (take-- <= 0)
+                {
+                    CleanExpired(context, expired);
+                    return true;
+                }
+
+                if (expired.TryGetValue(ticksSlice, out List<Slice> list) == false)
+                {
+                    list = new List<Slice> { keySlice };
+                }
+                else
+                {
+                    list.Add(keySlice);
+                }
+
+                expired[ticksSlice] = list;
+
+                if (items.ReadByKey(keySlice, out var reader) == false)
+                    continue;
+
+                var storeValue = reader.Read((int)ClusterStateMachine.CompareExchangeTable.Value, out var size);
+                using var result = new BlittableJsonReaderObject(storeValue, size, context);
+
+                if (HasExpiredMetadata(keySlice.ToString(), result, out long currentTicks) == false)
```

Comment:
Pass the Slice and materialize it only if needed inside the method?

Reply:
Aren't you doing double work here? `GetExpiredValues` already returns expired values and you are checking here again if they are expired?

Reply:
This is to handle update of `@expires` value
If the compare exchange `@expires` is updated there will be more than 1 entry for that compare exchange in the `CompareExchangeExpirationStorage`, so I have to check the value of compare exchange  `@expires` in the compare exchange table, then I can know the actual `@expires` value